<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Working with IMR data • RstoxData</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Working with IMR data">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RstoxData</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.5-9001</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/imrdata.html">Working with IMR data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/StoXProject/RstoxData/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Working with IMR data</h1>
                        <h4 data-toc-skip class="author">Edvin
Fuglebakk</h4>
            
            <h4 data-toc-skip class="date">2023-08-30</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/StoXProject/RstoxData/blob/master/vignettes/imrdata.Rmd" class="external-link"><code>vignettes/imrdata.Rmd</code></a></small>
      <div class="hidden name"><code>imrdata.Rmd</code></div>

    </div>

    
    
<p>This vignette introduces how to use RstoxData to work with data
collected at the Institute of Marine Research in Norway (IMR). It will
treat some common data formats and archiving systems at IMR, and may
refer internal resources that are not available outside the IMR
network.</p>
<div class="section level2">
<h2 id="documentation-resources">Documentation resources<a class="anchor" aria-label="anchor" href="#documentation-resources"></a>
</h2>
<p>Some key resources for data documentation at IMR is tabulated
below:</p>
<table class="table">
<colgroup>
<col width="14%">
<col width="40%">
<col width="44%">
</colgroup>
<thead><tr class="header">
<th>Link</th>
<th>Description</th>
<th>availability</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://havforskningsinstituttet.sharepoint.com/sites/Forskningsdata" class="external-link uri">https://havforskningsinstituttet.sharepoint.com/sites/Forskningsdata</a></td>
<td>Entry point for data documentation for all data at IMR</td>
<td>internal resource</td>
</tr>
<tr class="even">
<td><a href="https://datasetexplorer.hi.no/" class="external-link uri">https://datasetexplorer.hi.no/</a></td>
<td>portal for manual download of data</td>
<td>internal resource</td>
</tr>
<tr class="odd">
<td><a href="https://www.imr.no/formats/" class="external-link uri">https://www.imr.no/formats/</a></td>
<td>Formal description of data formats maintained at IMR by
HI-digital</td>
<td>public resource</td>
</tr>
<tr class="even">
<td><a href="https://referenceeditor.hi.no/apps/referenceeditor" class="external-link uri">https://referenceeditor.hi.no/apps/referenceeditor</a></td>
<td>Online system to inspect or download reference lists maintained at
IMR</td>
<td>internal resource</td>
</tr>
<tr class="odd">
<td><a href="https://hi.dkhosting.no/portal.aspx" class="external-link uri">https://hi.dkhosting.no/portal.aspx</a></td>
<td>Kvalietsportalen. QA-system at IMR. Contains some survey manuals,
sampling handbooks, etc.</td>
<td>internal resource</td>
</tr>
<tr class="even">
<td><a href="https://www.hi.no/hi/nettrapporter?query=&amp;serie=toktrapport" class="external-link uri">https://www.hi.no/hi/nettrapporter?query=&amp;serie=toktrapport</a></td>
<td>Cruise reports. Contain important information about the execution of
specific surveys</td>
<td>public resource</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="analysis-and-processing-with-other-stox-packages">Analysis and processing with other StoX packages<a class="anchor" aria-label="anchor" href="#analysis-and-processing-with-other-stox-packages"></a>
</h2>
<p>RstoxData provides support for reading and filtering various data
formats, but not much in terms of analysis or plotting. StoX and the
StoX R-packages provide facilities for many kinds of common analysis of
data at IMR. The StoX R-packages that provide functionality for StoX are
mostly also usable as independent libraries for R-scripting. For
fisheries independent analysis, consider the vignettes in RstoxFDA (<a href="https://stoxproject.github.io/RstoxFDA/" class="external-link uri">https://stoxproject.github.io/RstoxFDA/</a>), for analysis
of acoustic surveys or swept-area surveys consider RstoxBase (<a href="https://stoxproject.github.io/RstoxBase/" class="external-link uri">https://stoxproject.github.io/RstoxBase/</a>).</p>
</div>
<div class="section level2">
<h2 id="extendable-reference-lists">Extendable reference lists<a class="anchor" aria-label="anchor" href="#extendable-reference-lists"></a>
</h2>
<p>Most formats at IMR use extendable reference-list, which means that
most categorical variables may be subject to extensions at any time.
This is important to keep in mind, when considering re-usability of code
for analysis. Current reference lists may be viewed or downloaded with
referenceeditor (<a href="https://referenceeditor.hi.no/apps/referenceeditor" class="external-link uri">https://referenceeditor.hi.no/apps/referenceeditor</a>).
Historical codes that are no longer used are kept in the reference
lists, but their expiry will be indicated either in the field
‘deprecated’ or in the field ‘validTo’.</p>
</div>
<div class="section level2">
<h2 id="fish-and-invertebrate-data-nmdbiotic">Fish and invertebrate data (NMDbiotic)<a class="anchor" aria-label="anchor" href="#fish-and-invertebrate-data-nmdbiotic"></a>
</h2>
<p>The main source of biological data for fish and invertebrates at IMR
is the data archiving system NMDbiotic. NMDbiotic serves biological data
from a wide range of data collection programs, both fishery dependent
and fishery independent surveys. The data is archived as data sets,
typically corresponding to a cruise on a single vessel, or one delivery
form a sampling program. Each data set is versioned, so that it is
possible to obtain versions from previous dates if data has been
updated. The data format is documented on <a href="https://www.imr.no/formats/nmdbiotic/v3/" class="external-link uri">https://www.imr.no/formats/nmdbiotic/v3/</a> .</p>
<div class="section level3">
<h3 id="data-organization-and-access">Data organization and access<a class="anchor" aria-label="anchor" href="#data-organization-and-access"></a>
</h3>
<p>An overarching classification is provided by the assignment of
mission-types that distinguishes the data sets in terms of sampling
frame and methodology (<a href="https://referenceeditor.hi.no/apps/referenceeditor/v2/tables/missionType" class="external-link uri">https://referenceeditor.hi.no/apps/referenceeditor/v2/tables/missionType</a>).
Like all reference lists for NMDbiotic, the lists of mission-types is
extendable, but at present the missiontypes 4,5 and 21 are used for
fishery independent surveys. Most of the rest is various fishery
dependent data collection. Particularly fishery-dependent data sets are
often among several NMDbiotic data sets delivered from a cruise, which
again may be part of a cruise time series. The connection between data
sets, cruise and cruise time series is provided in the referencetable
cruiseseries (<a href="https://referenceeditor.hi.no/apps/referenceeditor/v2/tables/cruiseseries" class="external-link uri">https://referenceeditor.hi.no/apps/referenceeditor/v2/tables/cruiseseries</a>).</p>
<div class="section level4">
<h4 id="manual-download">Manual download<a class="anchor" aria-label="anchor" href="#manual-download"></a>
</h4>
<p>This classification by missiontypes, cruise or cruise-series is used
to organize the data in datasetexplorer (<a href="https://datasetexplorer.hi.no/" class="external-link uri">https://datasetexplorer.hi.no/</a>), which provide access to
manual download of data by data-set. It also contains tools to pull all
data from a year (see tools at <a href="https://datasetexplorer.hi.no/apps/datasetexplorer/v2/tools" class="external-link uri">https://datasetexplorer.hi.no/apps/datasetexplorer/v2/tools</a>).</p>
</div>
<div class="section level4">
<h4 id="api-access">API-access<a class="anchor" aria-label="anchor" href="#api-access"></a>
</h4>
<p>More customized extracts from NMDbiotic can be made programatically
by using the APIs (<a href="https://confluence.imr.no/display/API/API+Documentation" class="external-link uri">https://confluence.imr.no/display/API/API+Documentation</a>).
The Biotic API provides various services for downloading biotic-data
sets, and the cruise API allows lookup of relations between biotic
data-sets and cruises. There is currently no API for cruise-timeseries,
but a semi-automatic approach can be realized by downloading the
cruiseseries table and reading in that.</p>
</div>
<div class="section level4">
<h4 id="three-ways-to-get-the-data-you-need">Three ways to get the data you need<a class="anchor" aria-label="anchor" href="#three-ways-to-get-the-data-you-need"></a>
</h4>
<p>Depending on the volume of data to be analyzed, there are three ways
to obtain data.</p>
<ul>
<li>Locate the data sets in datasetexplorer and download them one by
one.</li>
<li>Use the tools in datasetexplorer to fetch all data for each year of
interest. Manual download by year, and then do subsequent filtering in
R.</li>
<li>Write scripts to use the cruiseseries table and APIs to download the
data sets you need. See Example 1: pull cruise series data using
API</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="data-format">Data Format<a class="anchor" aria-label="anchor" href="#data-format"></a>
</h3>
<p>NMDbiotic is a hierarchical data format with extendable reference
lists (see note in extendable reference list at the top of this
document), and is served as XML files from the API or download services.
It is constructed around the following hierarchy of records:</p>
<ul>
<li>mission
<ul>
<li>fishstation
<ul>
<li>catchsample
<ul>
<li>individual
<ul>
<li>agedetermination</li>
<li>tag</li>
<li>prey
<ul>
<li>preylengthfrequencytable</li>
<li>copepodedevstagefrequencytable</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<p>Not all levels are mandatory, but records cannot be recorded without
the parent record. For instance, an ‘individual’ is always linked to a
‘catchsample’, but not all ‘catchsample’ records have any individual
records. A record is identified by the fields listed as ‘keys’ in the
format documentation, along with its parent record. When read with
RstoxData these records are organized in data.tables corresponding to
each level, and keys from all records higher in the hierarchy is
included as well.</p>
<p>Analysts often prefer to merge the records they need into one big
table. Since the hierarchy branches off below individual, it is not
possible to construct one general table for analysis. Also note that key
parameters like age- and tag- parameters are in a many to one
relationship with individuals, so it is often necessary to do some
selective filtering of lower levels if these are to be used in a table
with one individual for each row.</p>
<p>RstoxData provides functions for fast parsing of XML and filtering of
hierarchical data, along with functions to prepare common combinations
of the tables above. See Example 3: Read and filter biotic data with
RstoxData.</p>
<div class="section level4">
<h4 id="conditional-reference-lists">Conditional reference lists<a class="anchor" aria-label="anchor" href="#conditional-reference-lists"></a>
</h4>
<p>NMDbiotic makes use of some conditional reference lists. Most
reference list simply list the options of values for a categorical
variable. Conditional reference lists have different options depending
on some other variable. For instance lists for specialstage depend on
the value of species and sex. In fact all of the conditional reference
list is conditioned on species, so in the format documentation for
NMDbiotic these fields are listed as linked to the taxa, and the exact
link to the reference list in question can be located by navigating the
reference list ‘taxa’ (<a href="https://referenceeditor.hi.no/apps/referenceeditor/v2/tables/taxa" class="external-link uri">https://referenceeditor.hi.no/apps/referenceeditor/v2/tables/taxa</a>).</p>
</div>
<div class="section level4">
<h4 id="reference-list-taxa">Reference list: Taxa<a class="anchor" aria-label="anchor" href="#reference-list-taxa"></a>
</h4>
<p>The taxa list is a particularly complex reference list, and one that
is particularly central to analysis of biotic data. Some features that
are not entirely self-explanatory and will be clarified here.</p>
<p>The records in taxa is identified by the field ‘tsn’. The name may
suggest that this is the ITIS taxonomical serialnumber, but this is not
universally the case. In some cases it is only an internal identifier,
and will not correctly identify the taxon in ITIS. The field ‘IMR’
indicates its status. Whenever IMR is TRUE the corresponding ‘tsn’ value
is not identifying the correct taxon in ITIS. The field ‘AphiaId’
identifies the taxa in the World Register of Marine Species (Worms),
when provided.</p>
<p>The sublist ‘synonyms’ provide alternative names for a taxon in
several languages. It has an option to mark which synonym is preferred,
but that flag is not interpreted by all clients (including at the moment
referenceeditor).</p>
<p>The sublist ‘lists’ contain references to various taxa-dependent
referencelists, such as specialstage, agingstructure, etc.</p>
<p>The sublist ‘restrictions’ contain parameters used for quality
checking when recording data: Max length, condition factor ranges, etc.
These are soft restrictions meant to prompt double checking of
measurments, and not meant to reflect absolute biological limits.</p>
</div>
</div>
<div class="section level3">
<h3 id="examples-nmdbiotic">Examples NMDbiotic<a class="anchor" aria-label="anchor" href="#examples-nmdbiotic"></a>
</h3>
<p>Below are provided some code examples for how to fetch, read and
filter NMDbioticdata. Support for fetching data is not within the scope
of RstoxData, but a suggestion for how to implement solutions for that
is included below. These suggestions make use of the NMDbiotic API, so
we will define some useful auxiliary functions for that first. Since
these depend on internal resources, this vignette does not actually
execute these functions, but their usage is explained in the examples
below.</p>
<p>We define one function to fetch a data set from NMDbiotic. We will
fetch the latest snapshop, rather than extract from the live database.
That means that data may be as day old, but it is faster and causes less
load on the server, which is not dimensioned for intensive use. The main
disadvantage with snapshots is that they are archived in the version of
biotic that was current at the time of snapshot creation. That means
that one may get some heterogeneity in formats. Since snapshots were
introduced after biotic 3, we can utilize backwards compatible parsers
in RstoxData to get around that issue. As long as all snapshots are some
variant of biotic 3, we can compile them in a file that declares the
namespace to be the latest biotic 3 format. In the functions below the
namespace is hard coded as biotic 3.1.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#' Download data set from NMD biotic</span></span>
<span><span class="co">#' @description </span></span>
<span><span class="co">#'  Downloads a data set from latest snapshot in NMD biotic v3, </span></span>
<span><span class="co">#'  and saves it to a file as XML (namespace: http://www.imr.no/formats/nmdbiotic/v3.1)</span></span>
<span><span class="co">#' @param targetfile filename to save dataset to</span></span>
<span><span class="co">#' @param path path to dataset.</span></span>
<span><span class="co">#' @param url address to NMD biotic v.3 (defaults to production address pr. Aug 2023)</span></span>
<span><span class="co">#' @param port port to use (defaults 8080)</span></span>
<span><span class="co">#' @param append if TRUE appends missions to existing file. If false writes to a new file and adds start and end missions-tag</span></span>
<span><span class="co">#' @examples </span></span>
<span><span class="co">#'  \dontrun{pullDataSet("test.xml", "Forskningsfartøy/2023/Johan Hjort_LDGJ/2023002005")}</span></span>
<span><span class="va">pullDataSet</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">targetfile</span>, <span class="va">path</span>, <span class="va">url</span><span class="op">=</span><span class="st">"http://tomcat7.imr.no"</span>, <span class="va">port</span><span class="op">=</span><span class="fl">8080</span>, <span class="va">append</span><span class="op">=</span><span class="cn">F</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">targetfile</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">append</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"File"</span>, <span class="va">targetfile</span>, <span class="st">"already exists"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">basepath</span> <span class="op">&lt;-</span> <span class="st">"apis/nmdapi/biotic/v3"</span></span>
<span>  <span class="va">datasetpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/URLencode.html" class="external-link">URLencode</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">basepath</span>, <span class="va">path</span>, <span class="st">"dataset"</span>, sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">snapshotpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/URLencode.html" class="external-link">URLencode</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">basepath</span>, <span class="va">path</span>, <span class="st">"snapshot"</span>, sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">snapshots</span> <span class="op">&lt;-</span> <span class="fu">httr</span><span class="fu">::</span><span class="fu"><a href="https://httr.r-lib.org/reference/GET.html" class="external-link">GET</a></span><span class="op">(</span><span class="va">url</span>, port<span class="op">=</span><span class="va">port</span>, path<span class="op">=</span><span class="va">snapshotpath</span>, query<span class="op">=</span><span class="st">"version=3.1"</span><span class="op">)</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">snapshots</span><span class="op">$</span><span class="va">status_code</span> <span class="op">==</span> <span class="fl">404</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Found no snapshots for dataset at:"</span>, <span class="va">snapshotpath</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">snapshotlist</span> <span class="op">&lt;-</span> <span class="fu">xml2</span><span class="fu">::</span><span class="fu">as_list</span><span class="op">(</span><span class="fu">xml2</span><span class="fu">::</span><span class="fu"><a href="http://xml2.r-lib.org/reference/read_xml.html" class="external-link">read_xml</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rawConversion.html" class="external-link">rawToChar</a></span><span class="op">(</span><span class="va">snapshots</span><span class="op">$</span><span class="va">content</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">snapshotvector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">snapshotlist</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="va">x</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">latestsnapsot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">snapshotvector</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="va">latestsnapshotpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/URLencode.html" class="external-link">URLencode</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">snapshotpath</span>, <span class="va">latestsnapsot</span>, sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">mission</span> <span class="op">&lt;-</span> <span class="fu">httr</span><span class="fu">::</span><span class="fu"><a href="https://httr.r-lib.org/reference/GET.html" class="external-link">GET</a></span><span class="op">(</span><span class="va">url</span>, port<span class="op">=</span><span class="va">port</span>, path<span class="op">=</span><span class="va">latestsnapshotpath</span>, query<span class="op">=</span><span class="st">"version=3.1"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">content</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rawConversion.html" class="external-link">rawToChar</a></span><span class="op">(</span><span class="va">mission</span><span class="op">$</span><span class="va">content</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">starttag</span> <span class="op">&lt;-</span> <span class="st">"&lt;missions xmlns=\"http://www.imr.no/formats/nmdbiotic/v3.1\"&gt;"</span></span>
<span>  <span class="va">endtag</span> <span class="op">&lt;-</span> <span class="st">"&lt;/missions&gt;"</span></span>
<span>  </span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="va">append</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/write.html" class="external-link">write</a></span><span class="op">(</span><span class="va">content</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, file<span class="op">=</span><span class="va">targetfile</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/write.html" class="external-link">write</a></span><span class="op">(</span><span class="va">starttag</span>, file<span class="op">=</span><span class="va">targetfile</span>, append <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">content</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">2</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">content</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/write.html" class="external-link">write</a></span><span class="op">(</span><span class="va">content</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, file<span class="op">=</span><span class="va">targetfile</span>, append <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="va">append</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/write.html" class="external-link">write</a></span><span class="op">(</span><span class="va">endtag</span>, file<span class="op">=</span><span class="va">targetfile</span>, append <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>    </span>
<span>  <span class="op">}</span></span>
<span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The data set path that needs to be passed to the function above is
constructed based on the following convention in terms on NMDbiotic
variables: * For cruise data:
/missiontypename/startyear/platformname/cruise * For other data,
e.g. fisheries dependent data:
/missiontypename/startyear/platformname/missionnumber</p>
<p>Note that the same variables may have different names in other
systems. E.g. the cruise API uses missiontype for what NMDbiotic calls
missiontypename.</p>
<p>The cruiseseries reference table (<a href="https://referenceeditor.hi.no/apps/referenceeditor/v2/tables/cruiseseries" class="external-link uri">https://referenceeditor.hi.no/apps/referenceeditor/v2/tables/cruiseseries</a>)
relates cruises to cruise series by ship name and cruise number. We can
use the cruise API to look up paths for these cruises that also work in
NMDbiotic:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Look up data paths for a cruise</span></span>
<span><span class="co">#' @param vessel name of vessel</span></span>
<span><span class="co">#' @param cruise cruise number</span></span>
<span><span class="co">#' @param url address to cruise API (defaults to production address pr. Aug 2023)</span></span>
<span><span class="co">#' @param port port to use (defaults 8080)</span></span>
<span><span class="co">#' @examples </span></span>
<span><span class="co">#'  \dontrun{getCruisePaths("Johan Hjort","2023002005")}</span></span>
<span><span class="va">getCruisePaths</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">vessel</span>, <span class="va">cruise</span>, <span class="va">url</span><span class="op">=</span><span class="st">"http://tomcat7.imr.no"</span>, <span class="va">port</span><span class="op">=</span><span class="fl">8080</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">basepath</span> <span class="op">&lt;-</span> <span class="st">"apis/nmdapi/cruise/v2"</span></span>
<span>  <span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>type<span class="op">=</span><span class="st">"findByCruise"</span>,</span>
<span>                shipname<span class="op">=</span><span class="va">vessel</span>,</span>
<span>                cruisenr<span class="op">=</span><span class="va">cruise</span><span class="op">)</span></span>
<span>  <span class="va">cruiseIds</span> <span class="op">&lt;-</span> <span class="fu">httr</span><span class="fu">::</span><span class="fu"><a href="https://httr.r-lib.org/reference/GET.html" class="external-link">GET</a></span><span class="op">(</span><span class="va">url</span>, port<span class="op">=</span><span class="va">port</span>, path<span class="op">=</span><span class="va">basepath</span>, query<span class="op">=</span><span class="va">query</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">cruiseIds</span><span class="op">$</span><span class="va">status_code</span> <span class="op">==</span> <span class="fl">404</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Found no data for cruise"</span>, <span class="va">cruise</span><span class="op">)</span>, <span class="st">"at"</span>, <span class="va">vessel</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">content</span> <span class="op">&lt;-</span> <span class="fu">xml2</span><span class="fu">::</span><span class="fu"><a href="http://xml2.r-lib.org/reference/read_xml.html" class="external-link">read_xml</a></span><span class="op">(</span><span class="va">cruiseIds</span><span class="op">$</span><span class="va">content</span><span class="op">)</span></span>
<span>  <span class="va">rows</span><span class="op">&lt;-</span><span class="fu">xml2</span><span class="fu">::</span><span class="fu"><a href="http://xml2.r-lib.org/reference/xml_children.html" class="external-link">xml_children</a></span><span class="op">(</span><span class="fu">xml2</span><span class="fu">::</span><span class="fu"><a href="http://xml2.r-lib.org/reference/xml_children.html" class="external-link">xml_child</a></span><span class="op">(</span><span class="va">content</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">s</span> <span class="kw">in</span> <span class="va">rows</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">name</span> <span class="op">&lt;-</span> <span class="fu">xml2</span><span class="fu">::</span><span class="fu"><a href="http://xml2.r-lib.org/reference/xml_attr.html" class="external-link">xml_attr</a></span><span class="op">(</span><span class="va">s</span>, <span class="st">"name"</span><span class="op">)</span></span>
<span>    <span class="va">result</span><span class="op">[[</span><span class="va">name</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">xml2</span><span class="fu">::</span><span class="fu"><a href="http://xml2.r-lib.org/reference/xml_text.html" class="external-link">xml_text</a></span><span class="op">(</span><span class="va">s</span>, trim <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span></span></code></pre></div>
<p>To get paths for data sets that are not organized as cruises we can
use the fact that these typically have a sensible organisation by
mission type and use NMDbiotic API to list out all available data
sets:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Find data set paths for all data sets within a set of missiontypes and years</span></span>
<span><span class="co">#' @param years vector fo years to find paths for</span></span>
<span><span class="co">#' @param missiontypenames vector of missiontypenames to find paths for</span></span>
<span><span class="co">#' @param url address to cruise API (defaults to production address pr. Aug 2023)</span></span>
<span><span class="co">#' @param port port to use (defaults 8080)</span></span>
<span><span class="co">#' @examples</span></span>
<span><span class="co">#'  \dontrun{getMissionTypePaths(c("Referanseflåten-Hav"), c(2021))}</span></span>
<span><span class="va">getMissionTypePaths</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">missiontypenames</span>, <span class="va">years</span>, <span class="va">url</span><span class="op">=</span><span class="st">"http://tomcat7.imr.no"</span>, <span class="va">port</span><span class="op">=</span><span class="fl">8080</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">basepath</span> <span class="op">&lt;-</span> <span class="st">"apis/nmdapi/biotic/v3"</span></span>
<span>  <span class="co">#datasetpath &lt;- URLencode(paste(basepath, path, "dataset", sep="/"))</span></span>
<span>  <span class="co">#snapshotpath &lt;- URLencode(paste(basepath, path, "snapshot", sep="/"))</span></span>
<span>  </span>
<span>  <span class="va">paths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">m</span> <span class="kw">in</span> <span class="va">missiontypenames</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">y</span> <span class="kw">in</span> <span class="va">years</span><span class="op">)</span><span class="op">{</span></span>
<span>      <span class="va">platformspath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/URLencode.html" class="external-link">URLencode</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">basepath</span>, <span class="va">m</span>, <span class="va">y</span>, sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span>
<span>      <span class="va">response</span> <span class="op">&lt;-</span> <span class="fu">httr</span><span class="fu">::</span><span class="fu"><a href="https://httr.r-lib.org/reference/GET.html" class="external-link">GET</a></span><span class="op">(</span><span class="va">url</span>, port<span class="op">=</span><span class="va">port</span>, path<span class="op">=</span><span class="va">platformspath</span><span class="op">)</span></span>
<span>      <span class="kw">if</span> <span class="op">(</span><span class="va">response</span><span class="op">$</span><span class="va">status_code</span> <span class="op">!=</span> <span class="fl">200</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Error in API call. Could not list platforms for "</span>, <span class="va">m</span>, <span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span>      <span class="op">}</span></span>
<span></span>
<span>      <span class="va">platforms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span>      <span class="va">content</span> <span class="op">&lt;-</span> <span class="fu">xml2</span><span class="fu">::</span><span class="fu"><a href="http://xml2.r-lib.org/reference/read_xml.html" class="external-link">read_xml</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rawConversion.html" class="external-link">rawToChar</a></span><span class="op">(</span><span class="va">response</span><span class="op">$</span><span class="va">content</span><span class="op">)</span><span class="op">)</span></span>
<span>      <span class="kw">for</span> <span class="op">(</span><span class="va">child</span> <span class="kw">in</span> <span class="fu">xml2</span><span class="fu">::</span><span class="fu"><a href="http://xml2.r-lib.org/reference/xml_children.html" class="external-link">xml_children</a></span><span class="op">(</span><span class="va">content</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="kw">for</span> <span class="op">(</span><span class="va">grandchild</span> <span class="kw">in</span> <span class="fu">xml2</span><span class="fu">::</span><span class="fu"><a href="http://xml2.r-lib.org/reference/xml_children.html" class="external-link">xml_children</a></span><span class="op">(</span><span class="va">child</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>          <span class="kw">if</span> <span class="op">(</span><span class="fu">xml2</span><span class="fu">::</span><span class="fu"><a href="http://xml2.r-lib.org/reference/xml_attr.html" class="external-link">xml_attr</a></span><span class="op">(</span><span class="va">grandchild</span>, <span class="st">"name"</span><span class="op">)</span><span class="op">==</span><span class="st">"platformpath"</span><span class="op">)</span><span class="op">{</span></span>
<span>            <span class="va">platforms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">platforms</span>, <span class="fu">xml2</span><span class="fu">::</span><span class="fu"><a href="http://xml2.r-lib.org/reference/xml_text.html" class="external-link">xml_text</a></span><span class="op">(</span><span class="va">grandchild</span><span class="op">)</span><span class="op">)</span></span>
<span>          <span class="op">}</span></span>
<span>        <span class="op">}</span></span>
<span>      <span class="op">}</span></span>
<span></span>
<span>      <span class="kw">for</span> <span class="op">(</span><span class="va">p</span> <span class="kw">in</span> <span class="va">platforms</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="va">deliverypath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/URLencode.html" class="external-link">URLencode</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">basepath</span>, <span class="va">m</span>, <span class="va">y</span>, <span class="va">p</span>, sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="va">response</span> <span class="op">&lt;-</span> <span class="fu">httr</span><span class="fu">::</span><span class="fu"><a href="https://httr.r-lib.org/reference/GET.html" class="external-link">GET</a></span><span class="op">(</span><span class="va">url</span>, port<span class="op">=</span><span class="va">port</span>, path<span class="op">=</span><span class="va">deliverypath</span><span class="op">)</span></span>
<span>        </span>
<span>        <span class="kw">if</span> <span class="op">(</span><span class="va">response</span><span class="op">$</span><span class="va">status_code</span> <span class="op">!=</span> <span class="fl">200</span><span class="op">)</span><span class="op">{</span></span>
<span>          <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Error in API call. Could not list missions for "</span>, <span class="va">m</span>, <span class="va">y</span>, <span class="va">p</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>        <span class="va">content</span> <span class="op">&lt;-</span> <span class="fu">xml2</span><span class="fu">::</span><span class="fu"><a href="http://xml2.r-lib.org/reference/read_xml.html" class="external-link">read_xml</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rawConversion.html" class="external-link">rawToChar</a></span><span class="op">(</span><span class="va">response</span><span class="op">$</span><span class="va">content</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="va">deliveries</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span>        <span class="kw">for</span> <span class="op">(</span><span class="va">child</span> <span class="kw">in</span> <span class="fu">xml2</span><span class="fu">::</span><span class="fu"><a href="http://xml2.r-lib.org/reference/xml_children.html" class="external-link">xml_children</a></span><span class="op">(</span><span class="va">content</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>          <span class="kw">for</span> <span class="op">(</span><span class="va">grandchild</span> <span class="kw">in</span> <span class="fu">xml2</span><span class="fu">::</span><span class="fu"><a href="http://xml2.r-lib.org/reference/xml_children.html" class="external-link">xml_children</a></span><span class="op">(</span><span class="va">child</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>            <span class="kw">if</span> <span class="op">(</span><span class="fu">xml2</span><span class="fu">::</span><span class="fu"><a href="http://xml2.r-lib.org/reference/xml_attr.html" class="external-link">xml_attr</a></span><span class="op">(</span><span class="va">grandchild</span>, <span class="st">"name"</span><span class="op">)</span><span class="op">==</span><span class="st">"delivery"</span><span class="op">)</span><span class="op">{</span></span>
<span>              <span class="va">deliveries</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">deliveries</span>, <span class="fu">xml2</span><span class="fu">::</span><span class="fu"><a href="http://xml2.r-lib.org/reference/xml_text.html" class="external-link">xml_text</a></span><span class="op">(</span><span class="va">grandchild</span><span class="op">)</span><span class="op">)</span></span>
<span>            <span class="op">}</span></span>
<span>          <span class="op">}</span></span>
<span>        <span class="op">}</span></span>
<span>        </span>
<span>        <span class="kw">for</span> <span class="op">(</span><span class="va">d</span> <span class="kw">in</span> <span class="va">deliveries</span><span class="op">)</span><span class="op">{</span></span>
<span>          <span class="va">paths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">paths</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">m</span>, <span class="va">y</span>, <span class="va">p</span>, <span class="va">d</span>, sep<span class="op">=</span><span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>      <span class="op">}</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">paths</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>We will also define a function that pulls data for a set of paths and
compiles them into one NMDbiotic v3.1 xml file:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Fetches biotic-data for a set of paths and compiles them into one data file</span></span>
<span><span class="co">#' @param targetfile name of xml file to be created</span></span>
<span><span class="co">#' @param paths vector of paths to fetch</span></span>
<span><span class="co">#' @param overwrite if TRUE any existing 'targetfile' will be overwritten.</span></span>
<span><span class="co">#' @param namespace namespace to declare for the file</span></span>
<span><span class="co">#'  \dontrun{pullBioticData("test.xml", c("Forskningsfartøy/2023/Johan Hjort_LDGJ/2023002005", "Forskningsfartøy/2023/Johan Hjort_LDGJ/2023002006"))}</span></span>
<span><span class="va">pullBioticData</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">targetfile</span>, <span class="va">paths</span>, <span class="va">overwrite</span><span class="op">=</span><span class="cn">F</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">targetfile</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">overwrite</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"File"</span>, <span class="va">targetfile</span>, <span class="st">"already exists."</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">starttag</span> <span class="op">&lt;-</span> <span class="st">"&lt;missions xmlns=\"http://www.imr.no/formats/nmdbiotic/v3.1\"&gt;"</span></span>
<span>  <span class="va">endtag</span> <span class="op">&lt;-</span> <span class="st">"&lt;/missions&gt;"</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/write.html" class="external-link">write</a></span><span class="op">(</span><span class="va">starttag</span>, file<span class="op">=</span><span class="va">targetfile</span><span class="op">)</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">p</span> <span class="kw">in</span> <span class="va">paths</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch</a></span><span class="op">(</span><span class="op">{</span>    </span>
<span>          <span class="fu">pullDataSet</span><span class="op">(</span><span class="va">targetfile</span>, <span class="va">p</span>, append <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>          <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"pulling biotic data for"</span>, <span class="va">p</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span>, error <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">e</span><span class="op">)</span><span class="op">{</span><span class="kw"><a href="https://rdrr.io/r/base/warning.html" class="external-link">warning</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Could not pull biotic data for"</span>, <span class="va">p</span>, <span class="va">e</span><span class="op">)</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/write.html" class="external-link">write</a></span><span class="op">(</span><span class="va">endtag</span>, file<span class="op">=</span><span class="va">targetfile</span>, append <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>So in order to get the data we want using the functions defined
above, we first need to obtain a set of paths. This is somewhat
different depending on whether data is assigned a cruise number
(typically fisheries independent data) or not (typically fisheries
dependent data).</p>
<div class="section level4">
<h4 id="example-1-pull-cruise-series-data-using-api">Example 1: pull cruise series data using API<a class="anchor" aria-label="anchor" href="#example-1-pull-cruise-series-data-using-api"></a>
</h4>
<p>A particular useful example is data that is assigned a cruise number,
and organised in cruise time series. We can get at those by downloading
an excel file from referenceeditor. Search for ‘cruiseseries’ at <a href="https://referenceeditor.hi.no/apps/referenceeditor/v2/tables" class="external-link uri">https://referenceeditor.hi.no/apps/referenceeditor/v2/tables</a>
and select download. Read in sheet 3 from that file with your favourite
excel-file parser and and filter out the cruise series you dont want
(‘Cruiseseries code’). The cruise series name and description can be
found either at sheet 1 or simply at referenceeditor (<a href="https://referenceeditor.hi.no/apps/referenceeditor/v2/tables/cruiseseries" class="external-link uri">https://referenceeditor.hi.no/apps/referenceeditor/v2/tables/cruiseseries</a>).
The parameters that need to be fed to getCruisePaths can be found in the
columns “Ship name” and “Cruisenr”. Obtain paths, and feed these paths
to pullBioticData. Finally, you may read in these data using
RstoxData::ReadBiotic. I have done this for the cruise series “Barents
Sea NOR-RUS ecosystem cruise in autumn” and downloaded the data for
years 2021-2022 in the a file called ‘ecosystemsurvey.xml’. The
resulting data is available as a data object in this package
(\code{).</p>
<p>Occasionally, the cruisetimeseries table contains errors. It is worth
plotting the survey area and time period for each year, and follow up if
any years contain activity outside the typical temporal or spatial range
of the survey.</p>
</div>
<div class="section level4">
<h4 id="example-2-pull-fisheries-dependent-data-using-api">Example 2: pull fisheries dependent data using API<a class="anchor" aria-label="anchor" href="#example-2-pull-fisheries-dependent-data-using-api"></a>
</h4>
<p>This works similarly as for the example pulling cruise series, but
since fisheries dependent data are organized by mission type and are not
assigned cruise numbers, you will obtain paths with getMissionTypePaths
in stead. For instance getMissionTypePaths(c(“Referanseflåten-Hav”,
“Referanseflåten-Kyst”), c(2020, 2021)) will compile a set of paths for
all data sets for both reference fleets in the years 2020 and 2021, and
getMissionTypePaths(“Fiskerisampling - Lotteri”, 2018:2023) will obtain
paths for the catch sampling lottery for the years 201-2023. As for
cruise data, you can pass this set of paths to pullBioticData and read
the downloaded file in with RstoxData::ReadBiotic.</p>
<p>Specify the data sources of interest based on the reference table <a href="https://referenceeditor.hi.no/apps/referenceeditor/v2/tables/missionType" class="external-link uri">https://referenceeditor.hi.no/apps/referenceeditor/v2/tables/missionType</a>
.</p>
</div>
<div class="section level4">
<h4 id="example-3-read-and-filter-biotic-data-with-rstoxdata-">Example 3: Read and filter biotic data with RstoxData.<a class="anchor" aria-label="anchor" href="#example-3-read-and-filter-biotic-data-with-rstoxdata-"></a>
</h4>
<p>RstoxData::ReadBiotic reads all the data in the biotic file. Usually
it is desirable to filter and reformat a bit for analysis. In StoX,
biotic data is typically converted to the StoxBiotic format. This is
convenient to do if you want to write scripts that use the same
functions that are available in the StoX user interface, and
particularly if you want to integrate with data provided in other
formats that StoX supports (such as the biotic format provided in the
ICES acoustic database). It cannot not however, represent all data that
is available in NMDbiotic. You may also opt to transform the biotic data
to flat 2-dimensional tables using RstoxData::PrepareNmdBioticTable.
This function support data extracts for all data that is available in
NMDbiotic 3.</p>
<div class="section level5">
<h5 id="working-with-stoxbiotic">Working with StoxBiotic<a class="anchor" aria-label="anchor" href="#working-with-stoxbiotic"></a>
</h5>
<p>StoxBiotic is a somewhat simpler data format that unifies the
different formats for biotic data that StoX supports. See the function
documentation for RstoxData::StoxBiotic for an explanation of the
format, and a description of how it relates to NMDbiotic and
ICESbiotic.</p>
<p>Working from the example data obtained above, we use the function
RstoxData::StoxBiotic to convert the data:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ecosystem_sb</span> <span class="op">&lt;-</span> <span class="fu">RstoxData</span><span class="fu">::</span><span class="fu"><a href="../reference/StoxBiotic.html">StoxBiotic</a></span><span class="op">(</span><span class="fu">RstoxData</span><span class="fu">::</span><span class="va"><a href="../reference/ecosystemsurvey_example.html">ecosystemsurvey_example</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in getBottomDepth(bottomdepthstart, bottomdepthstop): StoX: The</span></span>
<span><span class="co">#&gt; BottomDepth is calculated as the average of bottomdepthstart and</span></span>
<span><span class="co">#&gt; bottomdepthstop from NMDBiotic. bottomdepthstart is missing (NA) in 8,</span></span>
<span><span class="co">#&gt; bottomdepthstop is missing (NA) in 392 and both are missing in 8 out of 1220</span></span>
<span><span class="co">#&gt; stations.</span></span>
<span><span class="co">#&gt; Warning in getCatchFractionWeight(catchweight, catchproducttype, catchcategory, : StoX: There are catchproducttype that are not 1 (4, NA), but with non-missing catchweight. This results in missing CatchFractionWeight in StoxBiotic() for the following catchcategory of cruise 2021209:</span></span>
<span><span class="co">#&gt;  78089</span></span>
<span><span class="co">#&gt; Warning in getSampleWeight(lengthsampleweight, sampleproducttype, catchcategory, : StoX: There are sampleproducttype that are not 1 (NA), but with non-missing lengthsampleweight. This results in missing SampleWeight in StoxBiotic() for the following catchcategory of cruise 2021209:</span></span>
<span><span class="co">#&gt;  206587</span></span>
<span><span class="co">#&gt;  206588</span></span>
<span><span class="co">#&gt;  48739</span></span>
<span><span class="co">#&gt;  50524</span></span>
<span><span class="co">#&gt;  51611</span></span>
<span><span class="co">#&gt;  51612</span></span>
<span><span class="co">#&gt;  51671</span></span>
<span><span class="co">#&gt;  51678</span></span>
<span><span class="co">#&gt;  51701</span></span>
<span><span class="co">#&gt;  719327</span></span>
<span><span class="co">#&gt;  76336</span></span>
<span><span class="co">#&gt;  78088</span></span>
<span><span class="co">#&gt;  78089</span></span>
<span><span class="co">#&gt;  93294</span></span>
<span><span class="co">#&gt;  95109</span></span>
<span><span class="co">#&gt;  95110</span></span>
<span><span class="co">#&gt;  95167</span></span>
<span><span class="co">#&gt;  95168</span></span>
<span><span class="co">#&gt;  95496</span></span>
<span><span class="co">#&gt;  95533</span></span>
<span><span class="co">#&gt;  95542</span></span>
<span><span class="co">#&gt;  95573</span></span>
<span><span class="co">#&gt;  96967</span></span>
<span><span class="co">#&gt;  98276</span></span>
<span><span class="co">#&gt;  999301</span></span>
<span><span class="co">#&gt; Warning in getIndividualTotalLength(length, lengthmeasurement, catchcategory, : StoX: There are lengthmeasurement that are not 'E' (A, B, C, F, G, H, I, J, L, NA), but with non-missing length. This results in missing IndividualTotalLength in StoxBiotic() for the following catchcategory of cruise 2021209:</span></span>
<span><span class="co">#&gt;  161022</span></span>
<span><span class="co">#&gt;  161975</span></span>
<span><span class="co">#&gt;  161996</span></span>
<span><span class="co">#&gt;  162187</span></span>
<span><span class="co">#&gt;  162575</span></span>
<span><span class="co">#&gt;  162680</span></span>
<span><span class="co">#&gt;  164706</span></span>
<span><span class="co">#&gt;  165283</span></span>
<span><span class="co">#&gt;  165421</span></span>
<span><span class="co">#&gt;  171677</span></span>
<span><span class="co">#&gt;  172414</span></span>
<span><span class="co">#&gt;  206587</span></span>
<span><span class="co">#&gt;  206588</span></span>
<span><span class="co">#&gt;  564138</span></span>
<span><span class="co">#&gt;  644687</span></span>
<span><span class="co">#&gt;  719327</span></span>
<span><span class="co">#&gt;  78089</span></span>
<span><span class="co">#&gt;  82330</span></span>
<span><span class="co">#&gt;  82335</span></span>
<span><span class="co">#&gt;  82336</span></span>
<span><span class="co">#&gt;  82348</span></span>
<span><span class="co">#&gt;  82418</span></span>
<span><span class="co">#&gt;  82419</span></span>
<span><span class="co">#&gt;  82589</span></span>
<span><span class="co">#&gt;  82635</span></span>
<span><span class="co">#&gt;  82671</span></span>
<span><span class="co">#&gt;  93294</span></span>
<span><span class="co">#&gt;  95110</span></span>
<span><span class="co">#&gt;  95496</span></span>
<span><span class="co">#&gt;  95533</span></span>
<span><span class="co">#&gt;  95542</span></span>
<span><span class="co">#&gt;  96967</span></span>
<span><span class="co">#&gt;  97935</span></span>
<span><span class="co">#&gt;  98421</span></span>
<span><span class="co">#&gt;  98428</span></span></code></pre></div>
<p>Notice the warnings. StoxBiotic does only support one kind of length
measurement (total length), and only one kind of weight measurement
(Live weight / Round weight), so warnings are issued for all other kinds
of measurements. Some of these are data errors (missing producttype),
but some are legitimate data not supported by StoxBiotic (length
measurements A,B,…).</p>
<p>If our analysis is only concerned with catch weights, and not length
measurements and sample weights, we may address only the issue with
catchproducttype and move on. Commonly, we may also be interested in
only particular species, and the errors may not occur at all for the
species of interest. In that case, we may filter before converting to
StoX and see if the warnings disappear. See the section on hierarchical
filtering. Otherwise, we may have to do necessary data cleaning before
converting to StoX. E.g. Estimating live weights or total lengths.</p>
<p>Having inspected warnings, we may move on to use some standard StoX
functions in our analysis. You may familiarize yourself with RstoxBase
or RstoxFDA to learn which options are available.</p>
<p>If we want to integrate StoX functions into an analysis that also
needs some parameter not included in StoxBiotic, you may be able to do
that with the funciton RstoxData::addToStoxBiotic. For instance, you may
be interested in getting start and end positions of a station, even if
StoxBiotic only includes one position:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ecosystem_sb_pos</span> <span class="op">&lt;-</span> <span class="fu">RstoxData</span><span class="fu">::</span><span class="fu"><a href="../reference/AddToStoxBiotic.html">AddToStoxBiotic</a></span><span class="op">(</span><span class="va">ecosystem_sb</span>, <span class="fu">RstoxData</span><span class="fu">::</span><span class="va"><a href="../reference/ecosystemsurvey_example.html">ecosystemsurvey_example</a></span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"latitudestart"</span>, <span class="st">"latitudeend"</span>, <span class="st">"longitudestart"</span>, <span class="st">"longitudeend"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>You will find these extra position columns added to the Station table
of StoxBiotic. In StoxBiotic, haul and station is distinguished, and
both of them partially correspond to the fishstation table of NMDbiotic.
If you would rather have the positions appended to the Haul level, you
may specify that as well via the SplitTableAllocation argument:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ecosystem_sb_pos_haul</span> <span class="op">&lt;-</span> <span class="fu">RstoxData</span><span class="fu">::</span><span class="fu"><a href="../reference/AddToStoxBiotic.html">AddToStoxBiotic</a></span><span class="op">(</span><span class="va">ecosystem_sb</span>, <span class="fu">RstoxData</span><span class="fu">::</span><span class="va"><a href="../reference/ecosystemsurvey_example.html">ecosystemsurvey_example</a></span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"latitudestart"</span>, <span class="st">"latitudeend"</span>, <span class="st">"longitudestart"</span>, <span class="st">"longitudeend"</span><span class="op">)</span>, <span class="st">"Lowest"</span><span class="op">)</span></span></code></pre></div>
<p>This is particularly useful in the cases where there are several
hauls at some stations.</p>
<p>You may also further process StoxBiotic to a single flat
2-dimensional data table, with each row representing one individual:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ecosystem_sb_individual</span> <span class="op">&lt;-</span> <span class="fu">RstoxData</span><span class="fu">::</span><span class="fu"><a href="../reference/MergeStoxBiotic.html">MergeStoxBiotic</a></span><span class="op">(</span><span class="va">ecosystem_sb_pos_haul</span><span class="op">)</span></span></code></pre></div>
<p>or you may discard individuals completely and make and flatten the
table at some higher level in the hierarchy, e.g. Sample:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ecosystem_sb_sample</span> <span class="op">&lt;-</span> <span class="fu">RstoxData</span><span class="fu">::</span><span class="fu"><a href="../reference/MergeStoxBiotic.html">MergeStoxBiotic</a></span><span class="op">(</span><span class="va">ecosystem_sb_pos_haul</span>, <span class="st">"Sample"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="working-with-preparenmdbiotictable">Working with PrepareNmdBioticTable<a class="anchor" aria-label="anchor" href="#working-with-preparenmdbiotictable"></a>
</h5>
<p>If you need to access tables that are not supported by StoxBiotic, or
you want a close correspondance in variable naming and record
identification to the NMDbiotic format, you may prepare similar
2-dimensional tables with the function RstoxData::PrepareNmdBioticTable.
NMDbiotic is a bit more complex than StoxBiotic, in that the hierarchy
branches at certain points and there are some potential pitfalls to
correctly merging in ages or tagids to individuals for instance. The
function RstoxData::PrepareNmdBioticTable helps you get this right. You
may for instance create a table resembling the table of individuals we
created for StoxBiotic:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ecosystem_nmdb_individual</span> <span class="op">&lt;-</span> <span class="fu">RstoxData</span><span class="fu">::</span><span class="fu"><a href="../reference/PrepareNmdBioticTable.html">PrepareNmdBioticTable</a></span><span class="op">(</span><span class="fu">RstoxData</span><span class="fu">::</span><span class="va"><a href="../reference/ecosystemsurvey_example.html">ecosystemsurvey_example</a></span>, <span class="st">"Individual"</span><span class="op">)</span></span></code></pre></div>
<p>This table also contains parameters from the agrereading table and
tag table that are conceptually below the individual table in the
hierarchy, and which may contain more than one record for each
individual.</p>
<p>Records in NMDbiotic are identified via several key columns on some
tables, and via parent records. It is often convenient to extract record
identifiers in a single column. PrepareNmdBioticTable offers that with
the argument ‘addIds’:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ecosystem_nmdb_individual_ids</span> <span class="op">&lt;-</span> <span class="fu">RstoxData</span><span class="fu">::</span><span class="fu"><a href="../reference/PrepareNmdBioticTable.html">PrepareNmdBioticTable</a></span><span class="op">(</span><span class="fu">RstoxData</span><span class="fu">::</span><span class="va"><a href="../reference/ecosystemsurvey_example.html">ecosystemsurvey_example</a></span>, <span class="st">"Individual"</span>, <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Individual ids:"</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ecosystem_nmdb_individual_ids</span><span class="op">$</span><span class="va">Individual</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Individual ids: 285842"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Number of rows:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">ecosystem_nmdb_individual</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Number of rows: 285842"</span></span></code></pre></div>
<p>For more specialized analysis may consider extracting a table of
age-readings, for records where several readings are recorded for some
individuals. There are none in our example data set, but it can be
obtained with the argument TargetTable = “MultipleReadings”. A more
relevant extract for this example, that is not available in StoxBiotic
is to extract fish diet data (prey data):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ecosystem_nmdb_prey_ids</span> <span class="op">&lt;-</span> <span class="fu">RstoxData</span><span class="fu">::</span><span class="fu"><a href="../reference/PrepareNmdBioticTable.html">PrepareNmdBioticTable</a></span><span class="op">(</span><span class="fu">RstoxData</span><span class="fu">::</span><span class="va"><a href="../reference/ecosystemsurvey_example.html">ecosystemsurvey_example</a></span>, <span class="st">"Prey"</span>, <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p>These ectracts contain all data from parent tables, so this table has
a total of 183 columns.</p>
</div>
<div class="section level5">
<h5 id="hierarchical-filtering">Hierarchical filtering<a class="anchor" aria-label="anchor" href="#hierarchical-filtering"></a>
</h5>
<p>In some cases it is desirable to filter data before converting it to
one 2-dimensional table. For instance, you may want to remove species
that are not of interest before conversion, in order to reduce the
number of warnings that needs to be checked. RstoxData provides
functions for specifying these kinds of filters for all the hierarchical
data formats it supports. See for instance RstoxData::FilterBiotic or
RstoxData::FilterStoxBiotic.</p>
<p>We first specify the filters we want for each table in a lists of
lists that is names exactly like the BioticData. We will specify a
filter that retains all cod (commonname ‘torsk’) that is longer than 20
cm.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filterExpression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">filterExpression</span><span class="op">[[</span><span class="st">"ecosystemsurvey.xml"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">filterExpression</span><span class="op">[[</span><span class="st">"ecosystemsurvey.xml"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"catchsample"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"commonname == 'torsk'"</span><span class="op">)</span></span>
<span><span class="va">filterExpression</span><span class="op">[[</span><span class="st">"ecosystemsurvey.xml"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"individual"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"length &gt; .20"</span><span class="op">)</span></span></code></pre></div>
<p>We then apply the filter:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filteredData</span> <span class="op">&lt;-</span> <span class="fu">RstoxData</span><span class="fu">::</span><span class="fu"><a href="../reference/FilterBiotic.html">FilterBiotic</a></span><span class="op">(</span><span class="fu">RstoxData</span><span class="fu">::</span><span class="va"><a href="../reference/ecosystemsurvey_example.html">ecosystemsurvey_example</a></span>, <span class="va">filterExpression</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in FUN(X[[i]], ...): StoX: Filter on data from ecosystemsurvey.xml</span></span>
<span><span class="co">#&gt; returned empty tables "copepodedevstagefrequencytable"</span></span></code></pre></div>
<p>Whenever a filter returns some empty tables, we get a warning as
above. The filter has retained only catchsamples with cod:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">filteredData</span><span class="op">$</span><span class="va">ecosystemsurvey.xml</span><span class="op">$</span><span class="va">catchsample</span><span class="op">$</span><span class="va">commonname</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; torsk </span></span>
<span><span class="co">#&gt;  1086</span></span></code></pre></div>
<p>We have also removed any records lower in the hierarchy from the
catchsamples that have been removed:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Retained"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">filteredData</span><span class="op">$</span><span class="va">ecosystemsurvey.xml</span><span class="op">$</span><span class="va">individual</span><span class="op">)</span>, <span class="st">"out of"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="fu">RstoxData</span><span class="fu">::</span><span class="va"><a href="../reference/ecosystemsurvey_example.html">ecosystemsurvey_example</a></span><span class="op">$</span><span class="va">ecosystemsurvey.xml</span><span class="op">$</span><span class="va">individual</span><span class="op">)</span>, <span class="st">"individuals"</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="co">#&gt; [1] "Retained 6844 out of 285842 individuals"</span></span></code></pre></div>
<p>But we have not removed any stations:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Retained"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">filteredData</span><span class="op">$</span><span class="va">ecosystemsurvey.xml</span><span class="op">$</span><span class="va">fishstation</span><span class="op">)</span>, <span class="st">"out of"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="fu">RstoxData</span><span class="fu">::</span><span class="va"><a href="../reference/ecosystemsurvey_example.html">ecosystemsurvey_example</a></span><span class="op">$</span><span class="va">ecosystemsurvey.xml</span><span class="op">$</span><span class="va">fishstation</span><span class="op">)</span>, <span class="st">"stations"</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="co">#&gt; [1] "Retained 1220 out of 1220 stations"</span></span></code></pre></div>
<p>The option ‘FilterUpwards’ allows us to also remove records higher in
the hierarchy when that is desirable. This defaults to FALSE, as this
typically removes the ability to infer zero-catches:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filteredUp</span> <span class="op">&lt;-</span> <span class="fu">RstoxData</span><span class="fu">::</span><span class="fu"><a href="../reference/FilterBiotic.html">FilterBiotic</a></span><span class="op">(</span><span class="fu">RstoxData</span><span class="fu">::</span><span class="va"><a href="../reference/ecosystemsurvey_example.html">ecosystemsurvey_example</a></span>, <span class="va">filterExpression</span>, <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in FUN(X[[i]], ...): StoX: Filter on data from ecosystemsurvey.xml</span></span>
<span><span class="co">#&gt; returned empty tables "copepodedevstagefrequencytable"</span></span></code></pre></div>
<p>Note that we now have fewer catchsamples, because we have removed the
ones that have no fish longer than 20 cm:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">filteredUp</span><span class="op">$</span><span class="va">ecosystemsurvey.xml</span><span class="op">$</span><span class="va">catchsample</span><span class="op">$</span><span class="va">commonname</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; torsk </span></span>
<span><span class="co">#&gt;   479</span></span></code></pre></div>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Edvin Fuglebakk, Arne Johannes Holmin, Ibrahim Umar, Sindre Vatnehol, Anthony Hennessey, Espen Johnsen, Norwegian Institute of Marine Research.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

      </footer>
</div>






  </body>
</html>
